### DISCUSSION TITLE
APK Cloner

### DISCUSSION THREAD

**Discussion by SZRabinowitz** [REAndroid/APKEditor#26] [Ideas]
The user should specify the new package name and it should modify the APK to have the new package name. Maybe it should have a default value which is to update the last character by one (for example, com.example would become com.examplf).

This is confusing and requires correct logic. The java classes will stay the same, so activity names cant be modified. and any activity, service, receiver etc that begins with a `.` in the manifest will need to be edited to include the original package name, because those classes wont exist anymore once the package name is changed.

This can be useful for installing multiple versions of an app on your phone or for changing the package name in general (I can list a few use cases if you want).

**Comment by REAndroid** [comment]
I already done this on my private repo. I will merge it soon after advanced features are completed

**Reply by SZRabinowitz** [reply]
Wow that sounds really cool! Best of luck! Thank you so much for this amazing work!

**Comment by SZRabinowitz** [comment]
Now since apkeditor uses smali, will this feature be extended to smali classes? 

(in a way its easier because you dont need to pay attention what start with ., if you will anyway change all classes to match package)

**Reply by REAndroid** [reply]
My next goal is to integrate AI/GPT then rename classes will be minor thing.

**Comment by jaredcat** [comment]
Any updates on this?
I made this for my use case: https://github.com/jaredcat/apk-cloner
Ive used it specifically to make multiple Twitch apps that are patched with Revanced

**Reply by SZRabinowitz** [reply]
find and replace suited my needs so I never really pushed. I would love to contribute to this library sometimes but I stink in Java so I made a simple find/replace in python instead. 

What I love about this cli tool is that it works really nicely when used  with other programming languages. Especially the JSON.

**Comment by REAndroid** [comment]
**Notes:** 
 - Cloning apk is mainly about renaming package name, but we also need to change application name and icon.
 - Cloning/Renaming a package is a very complicated task in context of making "universal-cloner" that works for ALL apk. It is not a simple string search and replace task.
 - Our main goal is to make universal-cloner for at least 50-75% of existing apps.

**Factors challenging for renaming package:** 
 - Relative class name on AndroidManifest.xml activity/provider/service ... e.g ".ui.MainActivity"
 - Ambiguous provider, authority, permission ... having no pattern or resemblance with package name 
 - Ambiguous constants like url that should not be renamed
 - Hard coded strings on dex files: e.g `new File("/data/data/com.example/files/file");`
 - Usage on custom binary files, e.g. `global-metadata.bin` on unity game resources (we have editor library but not sure if it worth to make a part of ARSCLib ) 
 - Usage on resource strings, asset text, json, html ... files
 - Worst case: native libraries  

**Plans/Progress:**
 - Dex library on ARSCLib is complete, and we can edit classes, strings ... directly without disassembling dex files.
 - Simple renaming on manifest is complete.
 - Rename hard coded strings on dex files.
 - Similar to ProGuard, rename package/string based on user defined script.

FYI: _I've experimented and gave up AI assisted package renaming with GPT-4o, by creating string chat interface. Ignoring heavy token usage, this resulting worst to none and i don't know if better result could be achieved with custom trained data or model._
